### PR TITLE
🔧 Make type-checking opt-out

### PR DESCRIPTION
**Pull request by chrisjsewell**
Also update mypy and fix some failures

### CODE REVIEW

**Comment by chrisjsewell** [comment]
The nightly fail is independent of this PR

**Comment by ltalirz** [comment]
@sphuber is there a reason not to proceed with this PR?

I guess it's going to be quite prone to merge conflicts if left open for longer.

I see that besides type annotations, there are a few minor changes in behavior. @chrisjsewell is there any one of these that you think could be problematic?

**Comment by chrisjsewell** [comment]
It's good to go by me, just the one fix arising from merging the last commit.
I don't believe there should be any changes in behaviour

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -70,54 +70,138 @@ repos:
...
+                aiida/engine/processes/control.py|
+                aiida/engine/processes/functions.py|
+                aiida/engine/processes/ports.py|
+                aiida/manage/configuration/__init__.py|
+                aiida/manage/configuration/config.py|
+                aiida/manage/configuration/profile.py|
+                aiida/manage/configuration/settings.py|
+                aiida/manage/external/rmq/launcher.py|
+                aiida/manage/tests/main.py|
+                aiida/manage/tests/pytest_fixtures.py|
```
Comment (sphuber) [inline .pre-commit-config.yaml:122]: why are the lines
```
                aiida/orm/comments.py|
                aiida/orm/computers.py|
```
not removed? If they were properly typed before, should not exclude them now, should we?
Same for
```
                aiida/tools/graph/graph_traversers.py|
                aiida/tools/groups/paths.py|
```
further down

**Review by sphuber — COMMENTED** [review]
Thanks @chrisjsewell . Fine with the changes in principle, but just wondering how you created the list of files to exclude in the `.pre-commit-config.yaml`. I mentioned a few in a comment that used to be included (and so should be properly typed) but are now still part of the `exclude` list. Another example is the `aiida.engine` module which used to be included, but now there are many submodules that are excluded? What changed that they no longer pass type checking if you include them?

**Comment by chrisjsewell** [comment]
> but just wondering how you created the list of files to exclude

ran mypy it without the list, and then added all the files that were listed in the error log, or fixed them for most of the simplest issues

> What changed that they no longer pass type checking if you include them?

Mainly the removal of:

```toml
[[tool.mypy.overrides]]
module = 'aiida.*'
follow_imports = "skip"
```

i.e. the type checking now actually checks cross-package, whereas before it was relatively superficial

you'll see in the `pyproject.toml`, there are still a lot of aspects of mypy's strict type checking turned off, but this at least takes it one step further.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (chrisjsewell) [reply .pre-commit-config.yaml:122]: see https://github.com/aiidateam/aiida-core/pull/5811/#issuecomment-1338208930

**Review by sphuber — APPROVED** [review]

**Comment by sphuber** [comment]
Thanks @chrisjsewell . I have approved, but naturally would be good to fix the pre-commit before merging

**Pull request merged**